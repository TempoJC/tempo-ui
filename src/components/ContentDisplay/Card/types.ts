import type { ReactNode } from "react";
import { Color } from "@/types";

export type BasicCardProps = {
  children?: ReactNode | Array<ReactNode>;
  className?: string;
  testId?: string;
};

export type CardProps = BasicCardProps & {
  color?: Color;
  kind?: "default" | "outlined";
};

export type HeaderImageProps = BasicCardProps & {
  src?: string;
  width?: number;
};

export type ContentImageProps = BasicCardProps & {
  src?: string;
};
