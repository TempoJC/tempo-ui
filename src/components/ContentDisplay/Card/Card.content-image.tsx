import React from "react";
import cn from "classnames";
import { ContentImageProps } from "./types";

const ContentImage: React.FC<ContentImageProps> = ({ className, testId, src, children }) => (
  <div className={cn("tempo-card-content-image", className)} data-testid={testId}>
    <img src={src} />
    {children}
  </div>
);

export default ContentImage;
