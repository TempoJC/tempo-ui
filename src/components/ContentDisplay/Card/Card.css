.tempo-card-wrapper {
  display: flex;
  flex-direction: column;
}

.tempo-card-wrapper > * + * {
  margin-top: var(--size3);
}

.tempo-card-wrapper--outlined {
  border: 0.5px solid var(--color-token-light-border-1);
}

.tempo-card-wrapper--outlined.tempo-card-wrapper--dark {
  border: 0.5px solid var(--color-token-dark-border-1);
}

.tempo-card-header {
  display: flex;
}

.tempo-card-header > * + * {
  margin-left: var(--size3);
}

.tempo-card-header-image {
  position: relative;

  & img {
    display: block;
  }
}

.tempo-card-header-content {
  width: 100%;
  height: fit-content;
}

.tempo-card-content-image {
  position: relative;

  & img {
    width: 100%;
    display: block;
  }
}
