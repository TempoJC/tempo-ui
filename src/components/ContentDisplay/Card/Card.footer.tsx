import React from "react";
import cn from "classnames";
import { BasicCardProps } from "./types";

const Footer: React.FC<BasicCardProps> = ({ className, testId, children }) => (
  <div className={cn("tempo-card-footer", className)} data-testid={testId}>
    {children}
  </div>
);

export default Footer;
