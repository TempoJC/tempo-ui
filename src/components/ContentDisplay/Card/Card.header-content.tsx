import React from "react";
import cn from "classnames";
import { BasicCardProps } from "./types";

const HeaderContent: React.FC<BasicCardProps> = ({ className, testId, children }) => (
  <div className={cn("tempo-card-header-content", className)} data-testid={testId}>
    {children}
  </div>
);

export default HeaderContent;
