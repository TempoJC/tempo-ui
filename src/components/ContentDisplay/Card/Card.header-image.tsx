import React from "react";
import cn from "classnames";
import { HeaderImageProps } from "./types";

const HeaderImage: React.FC<HeaderImageProps> = ({ className, testId, src, children, width = 80 }) => (
  <div className={cn("tempo-card-header-image", className)} data-testid={testId}>
    <img style={{ width }} src={src} />
    {children}
  </div>
);

export default HeaderImage;
