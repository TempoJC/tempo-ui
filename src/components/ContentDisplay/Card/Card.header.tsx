import React from "react";
import cn from "classnames";
import { BasicCardProps } from "./types";

const Header: React.FC<BasicCardProps> = ({ className, testId, children }) => (
  <div className={cn("tempo-card-header", className)} data-testid={testId}>
    {children}
  </div>
);

export default Header;
