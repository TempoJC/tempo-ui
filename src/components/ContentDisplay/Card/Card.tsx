import React from "react";
import cn from "classnames";
import { Divider } from "@/components/Atoms/Divider";
import { CardProps } from "./types.js";
import "./Card.css";

/**
 * A card component.
 */
const Card: React.FC<CardProps> = ({ kind = "default", className, color = "light", testId, children }) => (
  <div
    className={cn("tempo-card-wrapper", className, {
      "tempo-card-wrapper--outlined": kind === "outlined",
      "tempo-card-wrapper--dark": color === "dark",
    })}
    data-testid={testId}
  >
    {children}
    {kind !== "outlined" && <Divider contrast="low" color={color} />}
  </div>
);

export default Card;
