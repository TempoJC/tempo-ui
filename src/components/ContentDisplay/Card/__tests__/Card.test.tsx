import { render, screen } from "@testing-library/react";
import Card from "../Card";
import Header from "../Card.header";
import HeaderImage from "../Card.header-image";
import HeaderContent from "../Card.header-content";
import ContentImage from "../Card.content-image";
import Footer from "../Card.footer";

describe("@tempo-ui/ContentDisplay/Card", () => {
  it("renders", () => {
    const { asFragment } = render(
      <Card className="customCardClass">
        <Header className="customHeaderClass">
          <HeaderImage src="https://picsum.photos/id/237/200/300" className="customHeaderImageClass" />
          <HeaderContent className="customHeaderContentClass">Lorem ipsum dolor sit amet.</HeaderContent>
        </Header>
        <ContentImage className="customContentImageClass" src="https://picsum.photos/id/1084/536/354" />
        <Footer className="customFooterClass">Footer content</Footer>
      </Card>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("sets dark color", () => {
    render(<Card color="dark" testId="card" />);
    expect(screen.getByTestId("card")).toHaveClass("tempo-card-wrapper--dark");
  });

  it("sets outlined type", () => {
    render(<Card kind="outlined" testId="card" />);
    expect(screen.getByTestId("card")).toHaveClass("tempo-card-wrapper--outlined");
  });

  describe("testId prop", () => {
    it("renders the Card and its children with the specified data-testid attributes", () => {
      render(
        <Card testId="card">
          <Header testId="card-header">
            <HeaderImage src="https://picsum.photos/id/237/200/300" testId="card-header-image" />
            <HeaderContent testId="card-header-content">Lorem ipsum dolor sit amet.</HeaderContent>
          </Header>
          <ContentImage testId="card-content-image" src="https://picsum.photos/id/1084/536/354" />
          <Footer testId="card-footer">Footer content</Footer>
        </Card>,
      );

      const cardElement = screen.getByTestId("card");
      expect(cardElement).toBeInTheDocument();
      expect(cardElement).toHaveAttribute("data-testid", "card");

      const headerElement = screen.getByTestId("card-header");
      expect(headerElement).toBeInTheDocument();
      expect(headerElement).toHaveAttribute("data-testid", "card-header");

      const headerImageElement = screen.getByTestId("card-header-image");
      expect(headerImageElement).toBeInTheDocument();
      expect(headerImageElement).toHaveAttribute("data-testid", "card-header-image");

      const headerContentElement = screen.getByTestId("card-header-content");
      expect(headerContentElement).toBeInTheDocument();
      expect(headerContentElement).toHaveAttribute("data-testid", "card-header-content");

      const contentImageElement = screen.getByTestId("card-content-image");
      expect(contentImageElement).toBeInTheDocument();
      expect(contentImageElement).toHaveAttribute("data-testid", "card-content-image");

      const footerElement = screen.getByTestId("card-footer");
      expect(footerElement).toBeInTheDocument();
      expect(footerElement).toHaveAttribute("data-testid", "card-footer");
    });
  });
});
