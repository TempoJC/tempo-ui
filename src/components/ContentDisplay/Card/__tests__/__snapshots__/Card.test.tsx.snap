// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`@tempo-ui/ContentDisplay/Card renders 1`] = `
<DocumentFragment>
  <div
    class="tempo-card-wrapper customCardClass"
  >
    <div
      class="tempo-card-header customHeaderClass"
    >
      <div
        class="tempo-card-header-image customHeaderImageClass"
      >
        <img
          src="https://picsum.photos/id/237/200/300"
          style="width: 80px;"
        />
      </div>
      <div
        class="tempo-card-header-content customHeaderContentClass"
      >
        Lorem ipsum dolor sit amet.
      </div>
    </div>
    <div
      class="tempo-card-content-image customContentImageClass"
    >
      <img
        src="https://picsum.photos/id/1084/536/354"
      />
    </div>
    <div
      class="tempo-card-footer customFooterClass"
    >
      Footer content
    </div>
    <div
      class="tempo-divider tempo-divider--light tempo-divider--horizontal tempo-divider--low"
    />
  </div>
</DocumentFragment>
`;
